{"cells":[{"cell_type":"code","execution_count":1,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["129 154 178 289 328 360 423 439 496 516 659 685 735 778 793 797 816 896 946 993 1006 1021 1065 1084 1120 1240 1243 1435 1459 1481 1606 1621 1833 1834 1867 1903 1909 1917 1952 2023 2035 2059 2084 2085 2197 2297 2426 2513 2583 2598 2692 2723 2727 2788 2798 2799 2932 2986 3046 3108 3146 3230 3364 3414 3450 3471 3506 3507 3549 3578 3661 3692 3745 3780 3899 3950 3974 3979 3989 4026 4142 4307 4309 4592 4622 4628 4662 4738 4746 4850 5058 5107 5187 5234 5235 5269 5391 5395 5468 5505 5563 5630 5633 5891 5900 5947 5964 5992 6059 6067 6160 6165 6222 6258 6413 6463 6475 6680 6691 6716 6796 6803 6815 6841 6878 6892 6920 6964 7000 7041 7049 7092 7205 7215 7236 7272 7322 7381 7398 7460 7498 7566 7577 7582 7632 7637 7658 7668 7719 7722 7804 7824 7885 7971 8099 8131 8211 8275 8296 8305 8325 8362 8363 8472 8482 8665 8701 8724 8733 8751 8831 8833 8847 8886 8920 8985 8986 8993\n","9005 8943 8849 8783 8748 8512 8310 8293 8260 8174 8076 7898 7862 7728 7713 7691 7665 7573 7518 7517 7487 7457 7449 7350 7315 7277 7192 7107 7045 7016 6992 6899 6886 6787 6517 6505 6473 6459 6381 6350 6329 6274 6252 6176 6065 6047 6043 6019 5852 5841 5831 5801 5770 5720 5694 5659 5654 5631 5609 5578 5462 5400 5335 5279 5192 5044 5038 4799 4727 4684 4640 4595 4573 4551 4535 4491 4434 4412 4354 4335 4316 4254 4140 4108 4022 4004 3919 3879 3745 3733 3683 3560 3489 3483 3436 3435 3385 3349 3297 3292 3279 3265 3262 3138 3042 3018 3010 2977 2968 2958 2955 2954 2848 2826 2821 2796 2791 2749 2741 2714 2568 2550 2545 2470 2256 2212 2169 2149 2116 2088 1980 1945 1859 1844 1841 1806 1774 1762 1662 1634 1618 1609 1562 1536 1508 1477 1452 1436 1359 1300 1248 1215 1154 1088 978 924 671 615 586 567 509 491 484 479 437 406 384 380 358 231 217 196 190 170 62 57 28 25 17\n"]}],"source":["# EX 1 - Bioinformatics Stronghold: Longest Increasing Subsequence - lgis\n","\n","\"\"\"\"\"\n","Given: A positive integer n≤10000 followed by a permutation π of length n.\n","\n","Return: A longest increasing subsequence of π, followed by a longest decreasing subsequence of π.\n","\"\"\"\"\"\n","\n","# I need to find the longest increasing and decreasing subsequences in a permutation.\n","# LIS = Longest Increasing Subsequence; LDS = Longest Decreasing Subsequence\n","\n","# Dynamic programming to find LIS\n","# This function: - keeps track of the length of the longest increasing subsequence (LIS) that ends at each position\n","# - For each element, it looks at all the previous elements to build the optimal subsequence\n","# - \"sequence\" is the input parameter of the function -> it's the permutation of numbers I am analyzing\n","def find_LIS(sequence):\n","    \n","    # I calculate the length of the sequence (ex: len([5, 1, 4, 2, 3]) = 5)\n","    # I assign this value to a variable called n\n","    # This line is useful because: - I need to know the length of the sequence to create support arrays of the right size; - I will use it in the following for loops to iterate over the sequence;\n","    n = len(sequence)\n","    \n","    # I create two arrays of length n - dp and prev:\n","\n","    # I create a dp (dynamic programming) array of length n where all elements are initialised to 1 -> It is initialised to 1 because each element alone is an increasing subsequence of length 1\n","    # Each position dp[i] will represent the length of LIS that ends with the element at position i\n","    dp = [1] * n\n","    \n","    # I create a prev array of length n where all elements are initialised to -1\n","    # Each position prev[i] stores the index of the previous element in LIS\n","    # I use -1 as the initial value to indicate ‘no previous item’\n","    prev = [-1] * n\n","    \n","    # For each position, I look at all the previous elements:\n","    # - Iteration on the sequence from the second element (index 1) -> because the first element alone is always an increasing subsequence of length 1 -> So I start from the \n","    # second element to try to extend the subsequences\n","    for i in range(1, n):\n","        \n","        # This loop iterates over all elements preceding the current element \"sequence[i]\"\n","        # I compare sequence[i] with all previous elements to see if it can extend an existing subsequence\n","        for j in range(i):\n","            \n","            # sequence[i] > sequence[j] -> It checks whether the current element is greater than the previous element -> This means that sequence[i] can be added at the end of a increasing subsequence\n","            # dp[j] + 1 > dp[i] -> It checks whether extending the subsequence ending in sequence[j] (of length dp[j]) with sequence[i] results in a subsequence longer than the current one\n","            # dp[i] contains the length of the LIS ending in sequence[i]\n","            if sequence[i] > sequence[j] and dp[j] + 1 > dp[i]:\n","                \n","                # If the previous condition is true, I update dp[i] with the length of the longest subsequence ending in sequence[j], plus 1 (to include sequence[i])\n","                # This update ensures that dp[i] stores the length of LIS ending in sequence[i].\n","                dp[i] = dp[j] + 1\n","                \n","                # I use prev to rebuild the actual sequence:\n","                # - I store the index j in prev[i]\n","                # - This will later allow me to reconstruct the actual subsequence from the last element\n","                prev[i] = j\n","\n","                # Recap about this for loop:\n","                # - It compares each sequence[i] element with all preceding elements sequence[j].\n","                # - It updates the length of LIS that ends in sequence[i] if it can be extended from a LIS that ends in sequence[j]\n","                # - It stores the index j in order to reconstruct the actual subsequence later on\n","    \n","    # Find the length of LIS (max_length):\n","    # I find the maximum value in the dp array\n","    # By taking the max() of all the values in dp, I'm finding the overall length of LIS\n","    # ex: if the final dp array is [1, 1, 2, 2, 3], then max_length will be set to 3, which is the length of LIS\n","    max_length = max(dp)\n","    \n","    # I find the index of the last element in LIS\n","    # dp.index(max_length) -> returns the index of the first occurrence of the max_length value in the dp array\n","    # ex:  if dp is [1, 1, 2, 2, 3], then dp.index(3) would return 4, since the value 3 (the maximum value) is at index 4 in the dp array.\n","    # This allows me to reconstruct the actual sequence of elements that make up LIS. By starting from the index of the last element (end_index) \n","    # and working backwards using the prev array, I can trace back through the sequence to find all the elements in LIS\n","    # I need to know the end_index because LIS can appear at different positions in the original sequence\n","    end_index = dp.index(max_length)\n","    \n","    # I reconstruct the actual elements of LIS from the information stored in the dp and prev arrays:\n","    \n","    # Empty list to store the elements of LIS in the correct order.\n","    result = []\n","    \n","    # Loop that continues as long as end_index is not equal to -1.\n","    # end_index variable -> previously set to the index of the last element in LIS\n","    while end_index != -1:\n","        \n","        # I append the element at the current end_index position in the original sequence to the result list\n","        # -> This adds the elements of LIS to the result list in reverse order\n","        result.append(sequence[end_index])\n","        \n","        # I set end_index to the value stored in prev[end_index] (prev array stores the index of the previous element in LIS for each index)\n","        # So, on LIS, I am starting from the last element and working my way towards the beginning\n","        end_index = prev[end_index]\n","    \n","    # After the loop completes, I have the elements of LIS in the result list, but in reverse order.\n","    # The [::-1] slice notation is used to reverse the order of the list, so that the elements are returned in the correct order.\n","    return result[::-1]  \n","\n","\n","# This function is similar as find_LIS, but reverses the comparison condition (I use \"<\" instead of \">\") in order to find the longest decreasing subsequence\n","# It takes the input \"sequence\" as a parameter\n","def find_LDS(sequence):\n","    \n","    # I save the length of the input sequence in the variable n -> in order to simplify array allocation and indexing\n","    n = len(sequence)\n","    \n","    # I initialize two arrays: dp and prev -> both initialized with 1 and -1 respectively\n","\n","    # dp stores the length of LDS ending at each index.\n","    dp = [1] * n\n","    \n","    # prev stores the index of the previous element in LDS ending at each index.\n","    prev = [-1] * n\n","    \n","    # Nested loops to iterate through the sequence -> I compare each element to all the previous elements\n","    for i in range(1, n):\n","        for j in range(i):\n","            \n","            # sequence[i] < sequence[j] -> checks if the current element is less than the previous element (which is the criteria for a decreasing subsequence)\n","            # dp[j] + 1 > dp[i] -> checks if extending the LDS ending at index j (of length dp[j]) with the current element sequence[i] results \n","            # in a longer decreasing subsequence than the current LDS ending at index i.\n","            if sequence[i] < sequence[j] and dp[j] + 1 > dp[i]:\n","                \n","                # If both conditions are true, I update dp[i] and prev[i] accordingly\n","                dp[i] = dp[j] + 1\n","                prev[i] = j\n","    \n","    # Length of the overall LDS\n","    max_length = max(dp)\n","    \n","    # Index of the last element in this \"overall LDS\"\n","    end_index = dp.index(max_length)\n","    \n","    # I reconstructs the actual elements of LDS by following the prev pointers backwards, starting from the last element\n","    # It's the same process as in the LIS case, just with the sequence being decreasing instead of increasing\n","    result = []\n","    while end_index != -1:\n","        result.append(sequence[end_index])\n","        end_index = prev[end_index]\n","    \n","    # I reverse the list to get the correct order\n","    return result[::-1]  \n","\n","# This function calls both functions above and returns both the sequences found\n","# It takes two parameters: n -> the length of the input permutation; permutation -> the input permutation, which is a list of integers\n","def LIS_LDS(n, permutation):\n","    \n","    # I call the find_LIS function, passing the permutation as an argument\n","    # The function returns the longest increasing subsequence, which is stored in the increasing variable\n","    increasing = find_LIS(permutation)\n","    \n","    # I call the find_LDS function, also passing the permutation as an argument\n","    # The function returns the longest decreasing subsequence, which is stored in the decreasing variable\n","    decreasing = find_LDS(permutation)\n","    \n","    # I return both the increasing and decreasing subsequences as a tuple\n","    return increasing, decreasing\n","\n","# This function takes two parameters: increasing -> the longest increasing subsequence, which is a list of integers;\n","# decreasing -> the longest decreasing subsequence, which is also a list of integers\n","def format_output(increasing, decreasing):\n","    \n","    # I format the output as requested -> I use an f-string (formatted string literal) to construct the final output\n","    # ' '.join(map(str, increasing)) -> takes the increasing list of integers and converts each element to a string using the map(str, increasing) call\n","    # The ' '.join() function then concatenates all the string representations of the elements, separated by a space character -> The result is a space-separated \n","    # string of the elements in the increasing list. \n","\n","    # '\\n' (newline character) -> acts as a separator between the increasing and decreasing subsequences in the final output\n","\n","    # ' '.join(map(str, decreasing)) -> this part is similar to the increasing part, but it operates on the decreasing list of integers\n","    # It converts each element to a string and joins them with space characters\n","    return f\"{' '.join(map(str, increasing))}\\n{' '.join(map(str, decreasing))}\"\n","\n","# This if statement checks if the current script is being run as the main program (as opposed to being imported as a module)\n","# The with statement ensures that the file is properly closed after the code inside the block is executed\n","if __name__ == \"__main__\":\n","    \n","    # I open the file in read mode (\"r\")\n","    with open(\"./bs-datasets/rosalind_lgis.txt\", \"r\") as file:\n","        \n","        # I read all the lines from the file and I store them in the \"lines\" list\n","        # Each element in \"lines\" will be a string representing a single line from the file\n","        lines = file.readlines()\n","        \n","        # The first line of the input file contains the length of the permutation, which is stored in the variable n\n","        # I use the strip() method to remove any leading or trailing whitespace characters from the line\n","        # I use the int() function to convert the string to an integer\n","        n = int(lines[0].strip())\n","        \n","        # The second line of the input file contains the permutation itself, which is a space-separated list of integers\n","        # lines[1].strip().split() -> splits the line into a list of strings, where each string represents a single integer in the permutation\n","        # I use the map(int, ...) function to convert each string to an integer\n","        # I use the list() function to convert the map object to a regular Python list, which is stored in the permutation variable\n","        permutation = list(map(int, lines[1].strip().split()))\n","    \n","    # I call the LIS_LDS function, passing the n (length of the permutation) and the permutation itself as arguments\n","    # The function returns the longest increasing subsequenc (LIS) e and the longest decreasing subsequence (LDS), which are stored in the increasing and decreasing variables, respectively\n","    increasing, decreasing = LIS_LDS(n, permutation)\n","    \n","    # I call the format_output function, passing the increasing and decreasing subsequences as arguments\n","    # The function returns the formatted output string, which is stored in the result variable\n","    result = format_output(increasing, decreasing)\n","    print(result)"]},{"cell_type":"code","execution_count":5,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["2 4 10 11 19 30 32 44 50 51 58 62 63 64 68 70 72 73 83 85 93 96 97 102 104 105 108 113 116 119 120 128 129 137 149 157 158 164 166 170 174 177 181 183 190 192 195 204 224 232 236 237 240 249 250 253 257 258 266 271 275 277 285 287 289 290 292 295 298 309 316 319 324 329 332 340 342 344\n"]}],"source":["# EX 2 - Bioinformatics Stronghold: Finding a Spliced Motif - sseq\n","\n","\"\"\"\"\"\n","Given: Two DNA strings s and t (each of length at most 1 kbp) in FASTA format.\n","\n","Return: One collection of indices of s in which the symbols of t appear as a subsequence of s. If multiple solutions exist, you may return any one.\n","\"\"\"\"\"\n","\n","# How I solve this:\n","# 1. I iterate through each character in t, and for each character, I search for that character in s (starting from the current position s_index)\n","# 2. I keep track of the indices where the characters from t are found in s, and return the list of these indices at the end\n","\n","\n","# I import SeqIO from the Biopython library, which makes it easy to read files in FASTA format\n","from Bio import SeqIO\n","\n","# This function searches for positions where the characters of string t appear as a subsequence in s\n","def indices_collection(s, t):\n","\n","    # I initialize an empty list \"indices\" to store the indices\n","    indices = []  \n","    \n","    # I start from the beginning of 's'\n","    s_index = 0  \n","    \n","    # I iterate through each character in 't'\n","    # On the first iteration, \"ch\" will be the first character in t, then the second character, and so on\n","    for ch in t:\n","        \n","        # I search for the current character \"ch\" in the string s, starting from the current position s_index\n","        # The loop continues as long as \"s_index\" is less than the length of \"s\" (meaning we haven't reached the end of s) and the character at the current \n","        # \"s_index\" position in \"s\" is not equal to the current \"ch\" from \"t\".\n","        while s_index < len(s) and s[s_index] != ch:\n","            \n","            # I increment the s_index variable by 1, moving the search position one character further in s\n","            s_index += 1\n","        \n","        # I check if the current character \"ch\" from \"t\" was found in \"s\"\n","        # If s_index is still less than the length of s, it means the character was found\n","        if s_index < len(s):  \n","            \n","            # If the character was found, I add the 1-based index (since the problem statement uses 1-based indexing) to the \"indices\" list\n","            indices.append(s_index + 1)  \n","            \n","            # I increment the s_index variable by 1 to move to the next position in s, so the next character from t can be searched for\n","            s_index += 1  \n","        \n","        else:\n","            \n","            # If the current character from t was not found in s, this else block is executed, and the loop is broken\n","            break  \n","    \n","    return indices\n","\n","if __name__ == \"__main__\":\n","    \n","    # I read the input using Biopython\n","    with open(\"./bs-datasets/rosalind_sseq.txt\", \"r\") as file:\n","        \n","        # I use SeqIO.parse() to read the file rosalind_sseq.txt in FASTA format. I obtain a list of records, where each record represents a sequence in the file\n","        # I extract the sequences as strings with str(record.seq)\n","        sequences = [str(record.seq) for record in SeqIO.parse(file, \"fasta\")]\n","\n","    # I assign the first sequence of the file to s and the second to t\n","    s, t = sequences[0], sequences[1]\n","\n","    # I find the subsequence indices\n","    result = indices_collection(s, t)\n","    \n","    # I print the subsequence indices in a readable format:\n","    # - map(str, result) -> map() is a function that applies a given function (in this case, str()) to each item of the result\n","    # - The str function converts each element to a string\n","    # - \" \".join() -> method that takes a sequence of strings and joins them into a single string, with a space (\" \") between each element. -> So, it will take the list of string-converted items and join them with spaces, creating one long string.\n","    print(\" \".join(map(str, result)))\n","\n"]},{"cell_type":"code","execution_count":3,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["TTAGGTCCGCCCTCGGGCATCCATTCGCTAACAGGAATCGTTGTGGCAACTTTGTCATACCTTGATACAGTTAAGCCCCAACAACCCGGGAGGCAACGTTTATATCGGGGTACTATATTGGCCGAACACACGCTTTTCACGGGGAGGGCGATTAGTGGGTAGAGGGGCATCAGTTTTTTGCACACGTAGCGCACTAACATATCTGACCAACGATAAGTTTGCCTATAGGAATCTGAAACTCCCCCATCACTTTTTCGCAATTGCCATGACTAGAACTCGCAGTGCCACAAAGGTAGCATCGGTTTTTCTCATTGTTTGGCTTATGACTGGGCCCGCTATATTGCTAAGTTCCAATTTCACAGAGGACCACTCGTGGACCGGTATTGACACCGAGATTGGATAACGATCTGGACGCATAGGGGTTCGCACCACGGTTTAGTTTATTGACCGTTTCAGAGATGTAGAAACACGGGTCAAACATATAGAAAGACTCTTTTTCCAACCGATTTGGTAAAATTACTTCTTAGGTTGGACATCGTTTCCGGGGAGCTGAAGGTT\n"]}],"source":["# EX 3 - Bioinformatics Stronghold: Finding a Shared Spliced Motif - lcsq\n","\n","\"\"\"\"\"\n","Given: Two DNA strings s and t (each having length at most 1 kbp) in FASTA format.\n","\n","Return: A longest common subsequence of s and t. (If more than one solution exists, you may return any one.)\n","\"\"\"\"\"\n","\n","# A subsequence is a sequence derived from another by removing some characters without changing their order (ex: \"ACTG\" is a common subsequence of \"AACCTTG\" and \"ACACTGTGA\")\n","# To find the longest common subsequence (LCS) between two strings \"s\" and \"t\", I use dynamic programming -> I break down the complex problem into simpler subproblems and store their solutions to avoid redundant calculations\n","# 1. I create a matrix to calculate the length of the LCS between all substrings of s and t; 2. I reconstruct the LCS from the matrix\n","\n","# I use Biopython -> SeqIO to read the sequences from the FASTA file\n","from Bio import SeqIO\n","\n","def LCS(s, t):\n","\n","    # I create a matrix to store LCS lengths:\n","    # I assign to variables \"m\" and \"n\", the length of the first string (s) and the length of the second string (t) respectively\n","    m, n = len(s), len(t)\n","\n","    # I create a matrix (m+1) x (n+1) dp where each cell dp[i][j] represents the length of the LCS between the first \"i\" characters of \"s\" and first \"j\" characters of \"t\"\n","    dp = [[0] * (n + 1) for _ in range(m + 1)]  \n","\n","    # I fill the matrix by comparing characters from both strings:\n","    # This loop runs through the rows of the dynamic matrix -> \"i\" represents the index of the string \"s\"; I start with 1 (not 0) to handle basic cases more easily;\n","    # I stop at m + 1 (length of s + 1) to cover the whole string\n","    for i in range(1, m + 1):\n","        \n","        # This cycle runs through the columns of the dynamic matrix -> \"j\" represents the index of the string \"t\"; I start from 1 and go up to n + 1;\n","        # For each \"i\", I scroll through all possible \"j\"\n","        for j in range(1, n + 1):\n","\n","            # When s[i - 1] is equal to t[j - 1], it means that I have found a common character\n","            if s[i - 1] == t[j - 1]:  \n","                \n","                # So I increase the length of the common subsequence by 1 -> that is, I take the value from the upper left diagonal (dp[i - 1][j - 1]) and add 1\n","                dp[i][j] = dp[i - 1][j - 1] + 1\n","            \n","            # When the characters are different, I take the maximum between: 1. Upper value (dp[i - 1][j]): ignoring the last letter of s;\n","            # 2. Left value (dp[i][j - 1]): ignoring the last letter of t\n","            else:  \n","                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])\n","\n","    # From the matrix, I reconstruct the LCS by following the corresponding characters and ignoring those not in common\n","    # I start from the bottom right of the matrix and trace the steps backwards by following the correspondences:\n","    \n","    # Empty list to store the common subsequence\n","    lcs = []\n","    \n","    # I set \"i\" and \"j\" to the last row and column of the matrix (original string lengths)\n","    i, j = m, n\n","\n","    # I continue until I reach the beginning of one of the two strings, so that I can explore the entire matrix by going back up\n","    while i > 0 and j > 0:\n","        \n","        # If the characters at positions i-1 and j-1 are identical..\n","        if s[i - 1] == t[j - 1]:  \n","            \n","            # .. I add the matching characters to the lcs list\n","            lcs.append(s[i - 1])\n","            \n","            # I move to the top left in the matrix\n","            i -= 1\n","            j -= 1\n","        \n","        # If the upper value is greater than or equal to the value on the left..\n","        elif dp[i - 1][j] >= dp[i][j - 1]:  \n","            \n","            # .. I move upwards (remove a character from s)\n","            i -= 1\n","        \n","        else:\n","            \n","            # Otherwise, I move to the left (remove a character from t)\n","            j -= 1\n","\n","    # I return the LCS as a string (I reverse the list to get the correct order)\n","    return ''.join(reversed(lcs))\n","\n","if __name__ == \"__main__\":\n","    \n","    # I read the sequences from the FASTA file using Biopython\n","    with open(\"./bs-datasets/rosalind_lcsq.txt\", \"r\") as file:\n","        \n","        # SeqIO.parse reads the FASTA file and returns the sequences as objects\n","        # With str(record.seq) I extract the sequence as a string\n","        sequences = [str(record.seq) for record in SeqIO.parse(file, \"fasta\")]\n","\n","    # The two sequences\n","    s, t = sequences[0], sequences[1]\n","\n","    # I return the LCS (longest common subsequence)\n","    result = LCS(s, t)\n","    print(result)\n"]},{"cell_type":"code","execution_count":15,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["300\n"]}],"source":["# EX 4 - Bioinformatics Stronghold: Edit Distance - edit\n","\n","\"\"\"\"\"\n","Given: Two protein strings s and t in FASTA format (each of length at most 1000 aa).\n","\n","Return: The edit distance d_E(s,t)\n","\"\"\"\"\"\n","\n","# I have to calculate the edit distance between two strings s and t.\n","# The edit distance is the minimum number of edit operations (substitution, insertion, or deletion of a single symbol) needed to transform s into t. \n","\n","# I use a dynamic programming approach to calculate the edit distance between the two input strings (s, t)\n","def edit_distance(s, t):\n","\n","    m, n = len(s), len(t)\n","\n","    # I create a two-dimensional table (a matrix) of size (m+1) x (n+1)\n","    # Each cell of the dp[i][j] array will represent the edit distance between the first \"i\" characters of \"s\" and the first \"j\" characters of \"t\"\n","    # I add 1 to ‘m’ and ‘n’ (i.e. I add one extra row and column) to handle the basic case where one of the strings is empty. \n","    # -> This allows me to calculate distances even when comparing against an empty string.\n","    dp = [[0] * (n + 1) for _ in range(m + 1)]\n","    \n","    # I fill the first row and column with sequential values: \n","    # dp[i][0]: This represents the edit distance between the first i characters of s and an empty string. The cost is \"i\" (all characters need to be deleted)\n","    # dp[0][j]: This represents the edit distance between an empty string and the first j characters of t. The cost is \"j\" (all characters need to be inserted)\n","    for i in range(m + 1):\n","        dp[i][0] = i\n","    for j in range(n + 1):\n","        dp[0][j] = j\n","    \n","    # I loop through each cell of the table, comparing the characters s[i-1] and t[j-1] (note the -1 because dp is 1-based while the strings are 0-based).\n","    # If the characters are the same (s[i-1] == t[j-1]) no edit is needed, so I copy the value from the previous diagonal cell dp[i-1][j-1].\n","    # If the characters are different, I calculate the minimum cost to make them the same using 3 possible operations: 1. Deletion: Remove the character s[i-1] → dp[i-1][j];\n","    # 2. Insertion: Add the character t[j-1] → dp[i][j-1]; 3. Substitution: Replace s[i-1] with t[j-1] → dp[i-1][j-1].\n","    # I add 1 to the minimum of these values to account for the cost of the current operation\n","    for i in range(1, m + 1):\n","        for j in range(1, n + 1):\n","            if s[i - 1] == t[j - 1]:\n","                dp[i][j] = dp[i - 1][j - 1]\n","            else:\n","                dp[i][j] = 1 + min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1])\n","    \n","    # I return the value stored in dp[m][n], which represents the minimum edit distance between the entire strings s and t\n","    return dp[m][n]\n","\n","# The problem asks for input in FASTA format, so the strings \"s\" and \"t\" are not in the first 2 lines of the file, but must be extracted correctly using a FASTA parser\n","# In FASTA format, each string is preceded by an identifier (ex: >Rosalind_39); Strings are distributed over several lines after the header.\n","# -> So I have to ignore the lines starting with > and concatenate the subsequent lines to get the sequences.\n","def parse_fasta(filename):\n","    \n","    # I open the input file (specified by file_path) and read all its lines into a list called \"lines\". Each line is stored as a separate string in the list.\n","    with open(filename, \"r\") as file:\n","        lines = file.readlines()\n","    \n","    sequences = {}\n","    current_id = None\n","    \n","    # I process the lines to extract the sequences and their IDs\n","    for line in lines:\n","        \n","        # If a line starts with >, it indicates a new sequence ID. I store this ID in the variable current_id and initialize an empty string for it in the dictionary \"sequences\".\n","        if line.startswith(\">\"):  \n","            current_id = line.strip()\n","            sequences[current_id] = \"\"\n","        \n","        else:\n","            \n","            # If a line does not start with >, it’s part of a sequence, so I append it to the string associated with the current ID in the dictionary.\n","            # At the end of this process, the sequences dictionary contains sequence IDs as keys and their corresponding sequences as values.\n","            sequences[current_id] += line.strip()\n","    \n","    # I extract only the sequences (ignoring their IDs) by taking the values from the \"sequences\" dictionary and converting them into a list\n","    # The result is a list of the two sequences that I’ll use for the edit distance calculation\n","    return list(sequences.values())\n","\n","if __name__ == \"__main__\":\n","    \n","    # I read input from the FASTA file\n","    filename = \"./bs-datasets/rosalind_edit.txt\"\n","    s, t = parse_fasta(filename)\n","    \n","    # I calculate the edit distance between s and t\n","    distance = edit_distance(s, t)\n","    print(distance)"]},{"cell_type":"code","execution_count":16,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["395\n","-GVMSSQQQPEWYDQIQMMDAPIFSYHRFMC--LHWI-NGNY-KGD---L--PLGDDPMREHPALFKPEVAGLWHVDSQQEFNQEFGIHIIDEEERLWGWQGPWYDHKWYRYMTMPCGSHCPFIWVGLHNMLLGQNKK-----QGL-M-------YHKLPTDFIWAGGWADVRSCNF--IEISQIQDGKNYFV-----AQLTLDAYWNNARWHLDAKQWQQRLDLKSLKYLHIVQPCEFAQGHDQIFSMVVTTFVFEVPLAKWSAGPFASGEAVCEDDCEYPMEPPNQNEPFISVD-FTVWCVNVDYNVQLN--PACEVESIICPYYMLQGMHYLFTGLWW----E---GVEWHGQGERDEVHLGPTLICNDEQDWSISFPIRWY-PYTWPIYMHYRTNFARTKETHIREENPQALYIQQWCMQDHKYQNTCPHCPGWPYGDLWLHSNWNPQPMPYIEDECNKMKGVFNGFEGYANYWSNPKYNGCGPFAFNIKWQQELIYIELIQDSEYANTANLIH-E-Q-----GYMIIWPISKYMRSSSNFPRRNRTR--PHIVIWVTCPKAWWAET--K-CRGFHRLPAHQYCTY---KGPRRKACHDVFVCT--ETGE-METPET-TTFGPNHRFHH------M-------MGRCEFTS--YA--PLM-HSG-KDS-R-L---GHEW-ACWYCFACTLTSDPHHTFYWATHGNENWGIVYWM---YFRNVCWDMEDSVHFWCQYDDTLWFNDTMHYM-TYQTAVPWACQTTQCI--NHEKHLTDLEY--Q-----PVNLEFEFCDQEQCFFRLFTMKRQGAQDPKYDLDRSLGDRYT--------VAGQGFY-----PDE\n","DDKWNCYQQPDIYEQIQ--IFPI-SLMDATCKSKMMIEIFSYTRADVCCLHWHNGDDPMVEHPALFK--------VDSHKYWTQHILIHIIGEEERLWGW--NNYRTLTY-YMTCPFIWTC-NGGYGLHNMLLGQTKKPDYRVRNLDMRQCTEVGDHKLPTDFIWAGG--TVADCNFDGIEISQIQDGKMYFVCPCEKLQLTLDAYWKSWQQRLD---LTRAAPPTSLKYLHEVQPCTFAQGHMQIQSMVVTTFVFEWPLAKCSAG-FCYGQYIFWRASGEAVCPPNQNEPFISVDNVDSIFFPVFLNVQHNSHPYPRTTQGVNVVTQHVQIHYLFEGLWWHEDIEHSPGVEWHGQGER-PVHLGPTLIFNDEQDWSI---IGMYFPIFMEHWNIYRTNFVQVK-HAWMQTKCNFSYGQDWCDQDHKYQNTCPHCPGWPYGDLWHFFHSNPQPMPYI---C-HFK--CN--VGYANYWSNPKYNGCGPFFFNIKWQQELIYDELIQDSEY--TANLIHEERQADHHKGYMCIWPILKYMRSSSNFPRKNRTRFYTRIRCPWTTKGAWWACTKYKAYRGFHRLPAGQYCTYRITWGWRRKCCHDVFVCTGIKAVEWHDSQSTNTTFGPNHRFHHEYVLIEMVVFYIVHCVRCEFTSYAYAWMQLMCRLGHEESCRPLVFNNICWSRCWFCFACTLTSDPHHTFYWATHGNEFWGPVYQMELASFYNVCWDMEDQVHFWRPYDDTLWFNDTMHRMPSYQCAGNKEPLIALVIEDDGEKHLTDLEYKHQYLIVGWVNREFEFCD--FPKELTFDPKTHGHQGLWTDLDRSL-DRQTRHLHNAMDEAGQGFYVHEALHDH\n"]}],"source":["# EX 5 - Bioinformatics Stronghold: Edit Distance Alignment - edta\n","\n","\"\"\"\"\"\n","Given: Two protein strings s and t in FASTA format (with each string having length at most 1000 aa).\n","\n","Return: The edit distance d_E(s,t) followed by two augmented strings s' and t' representing an optimal alignment of s and t.\n","\"\"\"\"\"\n","\n","def EditDistance(s, t):\n","    \n","    # 1. I initialize the DP table\n","    # I create a 2D DP table where dp[i][j] that will store the minimum edit distance to align the first i characters of s with the first j characters of t.\n","    \n","    # I store the lengths of the strings\n","    n, m = len(s), len(t)\n","    \n","    # I initialize a table of size (n+1) x (m+1)\n","    dp = [[0] * (m + 1) for _ in range(n + 1)]\n","\n","    # 2. Base cases (first row and column)\n","    # If one string is empty, the edit distance (the cost) is the length of the other string (all insertions or deletions).\n","    for i in range(n + 1):\n","        \n","        # I set the cost of deleting all characters in s\n","        dp[i][0] = i  \n","    \n","    for j in range(m + 1):\n","        \n","        # I set the cost of inserting all characters in t\n","        dp[0][j] = j  \n","\n","    # 3. I fill the DP table\n","    # I calculate the minimum cost for each cell considering insertion, deletion, and substitution:\n","    \n","    # I loop through all characters of s\n","    for i in range(1, n + 1):\n","        \n","        # I loop through all characters of t\n","        for j in range(1, m + 1):\n","            \n","            # No cost if the characters match, else -> Substitution cost (1) if the characters don't match\n","            cost = 0 if s[i - 1] == t[j - 1] else 1  \n","            \n","            # I calculate the minimum of the 3 operations: deletion, insertion, substitution:\n","            # Deletion\n","            dp[i][j] = min(dp[i - 1][j] + 1,     \n","                           \n","                           # Insertion\n","                           dp[i][j - 1] + 1,     \n","                           \n","                           # Substitution\n","                           dp[i - 1][j - 1] + cost)  \n","\n","    # 4. Traceback to reconstruct alignments\n","    # I use the DP table to trace back the optimal alignment of the strings.\n","    \n","    # I initialize two lists to store the aligned strings\n","    aligned_s, aligned_t = [], []\n","    \n","    # I start from the bottom-right corner of the DP table\n","    i, j = n, m\n","    \n","    # I trace back until I reach the top-left corner\n","    while i > 0 or j > 0:\n","        \n","        # If the current cell corresponds to a match or substitution..\n","        if i > 0 and j > 0 and dp[i][j] == dp[i - 1][j - 1] + (0 if s[i - 1] == t[j - 1] else 1):\n","            \n","             # .. I add the character from s\n","            aligned_s.append(s[i - 1])\n","            \n","            # .. I add the character from t\n","            aligned_t.append(t[j - 1])\n","            i -= 1\n","            j -= 1\n","        \n","        # If the current cell corresponds to a deletion in s..\n","        elif i > 0 and dp[i][j] == dp[i - 1][j] + 1:\n","            \n","            # .. I add the character from s\n","            aligned_s.append(s[i - 1])\n","            \n","            # .. I add a gap to t\n","            aligned_t.append('-')\n","            i -= 1\n","        \n","        else:  # j > 0 and dp[i][j] == dp[i][j - 1] + 1\n","            \n","            # If the current cell corresponds to an insertion in t:\n","\n","            # I add a gap to s\n","            aligned_s.append('-')\n","            \n","            # I add the character from t\n","            aligned_t.append(t[j - 1])\n","            j -= 1\n","\n","    # I reverse the alignments because I built them backwards during traceback:\n","    # -> I join the reversed list into a string\n","    aligned_s = ''.join(reversed(aligned_s))\n","    aligned_t = ''.join(reversed(aligned_t))\n","\n","    # I return the edit distance and the aligned strings\n","    return dp[n][m], aligned_s, aligned_t\n","\n","\n","def parse_fasta(filename):\n","    \n","    # I read the FASTA file and extract the sequences\n","    with open(filename, \"r\") as file:\n","        \n","        # I create a dictionary to store the sequences\n","        sequences = {}\n","        \n","        # I initialize a variable to store the current label\n","        current_label = None\n","        for line in file:\n","            \n","            # I remove any whitespace or newline characters\n","            line = line.strip()\n","            \n","            # If the line starts with \">\", it's a new sequence label\n","            if line.startswith(\">\"):\n","                \n","                # I store the label (without the \">\")\n","                current_label = line[1:]\n","                \n","                # I initialize an empty sequence for this label\n","                sequences[current_label] = \"\"\n","            \n","            else:\n","                \n","                # Otherwise, the line contains sequence data\n","                # -> I append the data to the current sequence\n","                sequences[current_label] += line\n","    \n","    # I return the sequences as a list\n","    return list(sequences.values())\n","\n","\n","\n","if __name__ == \"__main__\":\n","    \n","    # I parse the input from the FASTA file:\n","    # -> I read the sequences from the file\n","    sequences = parse_fasta(\"./bs-datasets/rosalind_edta.txt\")\n","    \n","    # I assign the first and second sequences to s and t\n","    s, t = sequences[0], sequences[1]\n","\n","    # I compute the edit distance and alignments\n","    distance, alignment_s, alignment_t = EditDistance(s, t)\n","\n","    # I format and print the result:\n","    \n","    # I print the edit distance\n","    print(distance)\n","    \n","    # I print the aligned version of s\n","    print(alignment_s)\n","    \n","    # I print the aligned version of t\n","    print(alignment_t)\n","\n"]},{"cell_type":"code","execution_count":18,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["111301358\n"]}],"source":["# EX 6 - Bioinformatics Stronghold: Counting Optimal Alignments - ctea\n","\n","\"\"\"\"\"\n","Given: Two protein strings s and t in FASTA format, each of length at most 1000 aa.\n","\n","Return: The total number of optimal alignments of s and t with respect to edit alignment score, modulo 134,217,727 (2^(27)-1).\n","\"\"\"\"\"\n","\n","def count_OptimalAlignments(s, t):\n","    \n","    # I use the modulo value given in the problem\n","    MOD = 134217727  \n","    \n","    # 1. I initialize DP tables (Dynamic Programming Table) -> I use it to calculate the edit distance between the two strings \n","    # dp[i][j]: Minimum edit distance for s[:i] and t[:j]\n","    n, m = len(s), len(t)\n","    dp = [[0] * (m + 1) for _ in range(n + 1)]\n","    \n","    # I maintain a second table \"count\" where count[i][j] stores the number of ways to achieve the edit distance at position (i, j).\n","    count = [[0] * (m + 1) for _ in range(n + 1)]\n","    \n","    # 2. Base cases -> I initialize the DP tables \"dp\" and \"count\" for scenarios where one of the two strings is empty\n","    # I loop through all possible prefixes of string s using the variable i, which ranges from 0 to n (inclusive)\n","    for i in range(n + 1):\n","        \n","        # Edit Distance (dp[i][0]): If string t is empty (has length 0), I need to delete all characters in s[:i] to align it with t\n","        # Thus, the edit distance for this alignment is \"i\" (the number of characters in s[:i])\n","        dp[i][0] = i  \n","        \n","        # count[i][0] -> There is only one way to delete all characters from s[:i] (by sequentially deleting each character). \n","        # Hence, I set count[i][0] = 1\n","        count[i][0] = 1  \n","    \n","    # I loop through all possible prefixes of string t using the variable j, which ranges from 0 to m (inclusive)\n","    for j in range(m + 1):\n","        \n","        # Edit Distance (dp[0][j]): If string s is empty (has length 0), I need to insert all characters in t[:j] to align it with s\n","        # Thus, the edit distance is j (the number of characters in t[:j])\n","        dp[0][j] = j  \n","        \n","        # count[0][j]: There is only one way to align s with t[:j] -> by sequentially inserting each character from t\n","        # So, I set count[0][j] = 1\n","        count[0][j] = 1  \n","    \n","    # 3. I fill DP tables:\n","    # I compute both the minimum edit distance (dp[i][j]) and the number of optimal alignments (count[i][j]) for all positions in the DP table.\n","    \n","    # I iterate through each prefix of s (up to length i) and t (up to length j), starting from 1.\n","    # The goal is to fill the DP table cell \"dp[i][j]\" and count table cell \"count[i][j]\".\n","    for i in range(1, n + 1):\n","        for j in range(1, m + 1):\n","            \n","            # Case 1: Match or Substitution -> I check if the last characters of the prefixes s[:i] and t[:j] are the same:\n","            # - If s[i-1] == t[j-1], no substitution is needed, and the cost of alignment is 0\n","            if s[i - 1] == t[j - 1]:\n","                cost = 0\n","            \n","            else: # - Otherwise, the characters differ, and the cost of substitution is 1\n","                cost = 1\n","            \n","            # I compute the minimum edit distance at dp[i][j] by considering the 3 possible operations:\n","            # Deletion ->  I delete the last character of s[:i] and align the rest of s[:i-1] with t[:j]. The cost is dp[i-1][j] + 1\n","            dp[i][j] = min(dp[i - 1][j] + 1,         \n","                           \n","                           # Insertion -> I insert the last character of t[:j] and align s[:i] with the rest of t[:j-1]. The cost is dp[i][j-1] + 1\n","                           dp[i][j - 1] + 1,         \n","                           \n","                           # Substitution/Match -> I align the last characters of s[:i] and t[:j] with the cost determined earlier (cost), and align the \n","                           # rest of s[:i-1] with t[:j-1]. The cost is dp[i-1][j-1] + cost.\n","                           dp[i - 1][j - 1] + cost)  \n","            \n","            # Case 2: I count Optimal Alignments (count[i][j])\n","            # -> For each valid move, I add the counts from the respective previous cells to the current cell while taking the modulo 134, 217, 727\n","            \n","            # I initialize the count for count[i][j] as 0\n","            count[i][j] = 0\n","            \n","            # Then, I add up the ways to achieve the minimum edit distance from each possible previous state:\n","\n","            # If the deletion operation is part of the optimal path (dp[i][j] == dp[i-1][j] + 1), I add all ways to align s[:i-1] with t[:j], which is count[i-1][j]\n","            if dp[i][j] == dp[i - 1][j] + 1:  \n","                count[i][j] += count[i - 1][j]\n","            \n","            # If the insertion operation is part of the optimal path (dp[i][j] == dp[i][j-1] + 1), I add all ways to align s[:i] with t[:j-1], which is count[i][j-1]\n","            if dp[i][j] == dp[i][j - 1] + 1:  \n","                count[i][j] += count[i][j - 1]\n","            \n","            # If the match/substitution operation is part of the optimal path (dp[i][j] == dp[i-1][j-1] + cost), I add all ways to align s[:i-1] with \n","            # t[:j-1], which is count[i-1][j-1].\n","            if dp[i][j] == dp[i - 1][j - 1] + cost:  \n","                count[i][j] += count[i - 1][j - 1]\n","            \n","            # To avoid integer overflow, I take the result modulo 134, 217, 727 at each step.\n","            count[i][j] %= MOD\n","    \n","    # 4.After filling the DP table, the total number of optimal alignments between s and t is stored in count[n][m]. I return this value.\n","    return count[n][m]\n","\n","\n","def parse_fasta(filename):\n","    \n","    # I read the FASTA file and extract the sequences\n","    with open(filename, \"r\") as file:\n","        \n","        # I create a dictionary to store the sequences\n","        sequences = {}  \n","        \n","        # I initialize a variable to store the current label\n","        current_label = None  \n","        for line in file:\n","            \n","            # I remove any whitespace or newline characters\n","            line = line.strip()  \n","            if line.startswith(\">\"):\n","                \n","                # If the line starts with \">\", it's a new sequence label:\n","                \n","                # I store the label (without the \">\")\n","                current_label = line[1:]  \n","                \n","                # I initialize an empty sequence for this label\n","                sequences[current_label] = \"\"  \n","            \n","            else:\n","                \n","                # Otherwise, the line contains sequence data:\n","                \n","                # I append the data to the current sequence\n","                sequences[current_label] += line  \n","    \n","    # I return the sequences as a list\n","    return list(sequences.values())  \n","\n","\n","\n","if __name__ == \"__main__\":\n","    \n","    # I parse the input from the FASTA file\n","    # -> I read the sequences from the file\n","    sequences = parse_fasta(\"./bs-datasets/rosalind_ctea.txt\")  \n","    \n","    # I assign the first and second sequences to s and t\n","    s, t = sequences[0], sequences[1]  \n","\n","    # I compute the number of optimal alignments\n","    total_alignments = count_OptimalAlignments(s, t)\n","\n","    # I print the number of optimal alignments modulo 134,217,727 (the result)\n","    print(total_alignments)  \n"]},{"cell_type":"code","execution_count":19,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["2629\n"]}],"source":["# EX 7 - Bioinformatics Stronghold: Global Alignment with Scoring Matrix - glob\n","\n","\"\"\"\"\"\n","Given: Two protein strings s and t in FASTA format (each of length at most 1000 aa).\n","\n","Return: The maximum alignment score between s and t. Use:\n","- The BLOSUM62 scoring matrix.\n","- Linear gap penalty equal to 5 (i.e., a cost of -5 is assessed for each gap symbol).\n","\"\"\"\"\"\n","\n","# I import the \"product\" function from the \"itertools\" module. \n","# \"product\" is used to generate all pairs of indices between two ranges (in this case, between the lengths of the two sequences s and t)\n","from itertools import product\n","\n","# I import the substitution_matrices module from BioPython. \n","# It allows you to load predefined scoring matrices for sequence alignment, like BLOSUM62, which is commonly used for protein sequence alignment.\n","from Bio.Align import substitution_matrices\n","\n","# I load the BLOSUM62 substitution matrix using the load function. The matrix contains the scores for aligning different amino acids. \n","# If two amino acids match, you get a positive score; if they mismatch, you get a negative score.\n","blosum62 = substitution_matrices.load(\"BLOSUM62\")\n","\n","\n","# This function takes two strings s and t (representing protein sequences) and returns the maximum alignment score between them using dynamic programming.\n","def MaxAlignment(s, t):\n","    \n","    # I calculate the lengths of the two input strings s and t, storing them in sl and tl respectively.\n","    sl, tl = len(s), len(t)\n","    \n","    # I initialize a dictionary \"m\" with a single key-value pair, where (0, 0) corresponds to the starting point of the alignment grid\n","    # The value (0, None) indicates a score of 0 and None as the previous alignment position\n","    m = {(0, 0): (0, None)}\n","    \n","    # I initialize the boundary conditions for the dynamic programming grid:\n","\n","    # I set the scores for the first column of the grid (aligning s to gaps in t). Each score is calculated as i * -5 (a gap penalty of -5 per insertion).\n","    m.update({((i, 0), (i * - 5, (i - 1, 0))) for i in range(1, sl + 1)})\n","    \n","    # I set the scores for the first row of the grid (aligning t to gaps in s), also applying a gap penalty of -5.\n","    m.update({((0, i), (i * - 5, (0, i - 1))) for i in range(1, tl + 1)})\n","    \n","    # I iterate over all possible pairs of positions (i, j) in the sequences s and t, excluding the first row and first column, which were already initialized.\n","    for i, j in product(range(1, sl + 1), range(1, tl + 1)):\n","        \n","        # I retrieve the substitution score for aligning the amino acids s[i - 1] and t[j - 1] from the BLOSUM62 matrix.\n","        cost = blosum62.get((s[i - 1], t[j - 1]))\n","        \n","        # If the score for the pair (s[i-1], t[j-1]) is not found (i.e., it's None)..\n","        if cost == None:\n","            # .. I try to get the score for the reverse pair (t[j-1], s[i-1]), since the matrix might be asymmetric.\n","            cost = blosum62.get((t[j - 1], s[i - 1]))\n","        \n","        # I calculate the possible scores for the 3 potential moves in the dynamic programming grid:\n","        \n","        # d -> diagonal move, meaning both s[i-1] and t[j-1] are aligned, so I add the substitution score cost to the value from the diagonal cell (i-1, j-1).\n","        d = m[(i - 1, j - 1)][0] + cost\n","        \n","        # l -> left move, meaning s[i-1] is aligned to a gap in t, so I subtract 5 for the gap penalty.\n","        l = m[(i - 1, j)][0] - 5\n","        \n","        # u -> up move, meaning t[j-1] is aligned to a gap in s, with a gap penalty of 5.\n","        u = m[(i, j - 1)][0] - 5\n","        \n","        # b stores the best score from the 3 possibilities.\n","        b = max(d, l, u)\n","        \n","        # I update the dynamic programming grid \"m\" by storing the best score \"b\" and the position of the move that resulted in this score \n","        # (the direction from where the value came).\n","        if d == b:\n","            m[(i, j)] = (b, (i - 1, j - 1))\n","        elif l == b:\n","            m[(i, j)] = (b, (i - 1, j))\n","        elif u == b:\n","            m[(i, j)] = (b, (i, j - 1))\n","            \n","    # I return the maximum alignment score stored at the bottom-right cell of the grid (i, j), which is the last cell processed\n","    # The score is converted to an integer to avoid any decimal part.\n","    return int(m[(i, j)][0])\n","\n","    \n","if __name__ == '__main__':\n","    \n","    # I initialize an empty list \"results\" to store the sequences read from the input file and an empty string \"protein\" to build each protein sequence line by line\n","    results = []\n","    protein = \"\"\n","    \n","    # I read the input file line by line\n","    with open('./bs-datasets/rosalind_glob.txt') as file:\n","        for line in file:\n","            line = line.rstrip()\n","\n","            # If a line starts with \">\", it marks the start of a new sequence, and the previous sequence (stored in \"protein\") is added to the results list.\n","            if line.startswith(\">\"):\n","                if(protein != \"\"):\n","                    results.append(protein)\n","                    protein = \"\"\n","            else:\n","                # Otherwise, the line is part of a sequence and is appended to \"protein\"\n","                protein += line\n","                \n","    # I append the last sequence (in case it’s not followed by a \">\" marker) to the results list.\n","    results.append(protein)\n","    \n","    # I call the function with the first two sequences from the results list, and the resulting maximum alignment score is printed.\n","    print (MaxAlignment(results[0], results[1]))\n"]},{"cell_type":"code","execution_count":2,"metadata":{"executionInfo":{"elapsed":403,"status":"ok","timestamp":1732181925083,"user":{"displayName":"Marco Zecchini","userId":"04080461015443912025"},"user_tz":-60},"id":"NZLKEFGqpkza"},"outputs":[],"source":["from random import randint, choice"]},{"cell_type":"markdown","metadata":{"id":"7kMDITdMpkzc"},"source":["# Map Reduce\n","\n","In the part of the assignment you are requested to use Map Reduce paradigm to solve the following exercises.\n","\n","**NOTE THAT**: **A solution that does not use map reduce is not valid!**"]},{"cell_type":"markdown","metadata":{"id":"3UvFKq5Dpkzd"},"source":["# Exercise 1\n","\n","You have a list of dictionaries, each representing a student with the following properties: a name and an array of test scores. Your task is to use map, filter, and reduce to calculate the average test score for each student, and then return a list of dictionaries containing only the students whose average score is above 90."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"XnZxeOdtpkze"},"outputs":[],"source":["students = [\n","    {\"name\": \"Alice\", \"scores\": [95, 92, 88, 100]},\n","    {\"name\": \"Bob\", \"scores\": [78, 81, 85, 80]},\n","    {\"name\": \"Charlie\", \"scores\": [99, 91, 94, 96]},\n","    {\"name\": \"Diana\", \"scores\": [85, 87, 89, 83]}\n","]"]},{"cell_type":"markdown","metadata":{"id":"WopOVxHSpkzf"},"source":["Use `map`, `reduce` and `filter` that produce an output like:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"JGYCqweYpkzf","outputId":"89230ab4-a15a-4ae5-84a3-9eb13573a4a0"},"outputs":[{"data":{"text/plain":["[{'name': 'Alice', 'average_score': 93.75},\n"," {'name': 'Charlie', 'average_score': 95.0}]"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["[\n","    {\"name\": \"Alice\", \"average_score\": 93.75},\n","    {\"name\": \"Charlie\", \"average_score\": 95.0}\n","]"]},{"cell_type":"markdown","metadata":{"id":"TG5h3pN1pkzg"},"source":["### Test\n","Test your solution using the dataset generated by the following function."]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":285,"status":"ok","timestamp":1732181928585,"user":{"displayName":"Marco Zecchini","userId":"04080461015443912025"},"user_tz":-60},"id":"3ZB3sFT_pkzh","outputId":"cf486e14-4d6f-455d-8f69-19096dfe9831"},"outputs":[],"source":["def generate_random_student_dataset(num_students=50):\n","    names = [f\"Student {i}\" for i in range(1, num_students + 1)]\n","    dataset = [\n","        {\n","            \"name\": name,\n","            \"scores\": [randint(50, 100) for _ in range(randint(3, 6))]  # Random scores between 50 and 100\n","        }\n","        for name in names\n","    ]\n","    return dataset\n","\n","random_student_dataset = generate_random_student_dataset(50)\n","random_student_dataset[:3]"]},{"cell_type":"code","execution_count":28,"metadata":{"id":"MHi3FLeWpkzi"},"outputs":[{"name":"stdout","output_type":"stream","text":["{'name': 'Student 11', 'average_score': 91.0} \n","\n","{'name': 'Student 12', 'average_score': 90.8} \n","\n","{'name': 'Student 49', 'average_score': 90.67} \n","\n"]}],"source":["# your code goes here\n","\n","# I import the randint function from the random module, which allows me to generate random integers\n","# I'll use this to create random test scores for each student.\n","from random import randint\n","\n","# I import the reduce function from the functools module.\n","# reduce allows me to perform a rolling computation on a sequence of values (in this case, the test scores) and accumulate them into a single result (the total score)\n","from functools import reduce\n","\n","# This function takes one argument (student) which is a dictionary containing the student's name and scores.\n","# It calculates the average score for each student based on their test scores.\n","def AverageCalculation(student):\n","    \n","    # I use reduce to sum the test scores -> reduce takes a function and a sequence, and applies the function cumulatively to the items in the sequence.\n","    # In this case, the lambda function lambda x, y: x + y adds the values x and y together. \n","    total_score = reduce(lambda x, y: x + y, student['scores'])  \n","    \n","    # To calculate the average, I need to know how many scores are in the list, so I can divide the total score by the correct number.\n","    # so -> I get the number of test scores for the student by calculating the length of the \"scores\" list.\n","    scores_number = len(student['scores'])  \n","    \n","    # Formula for calculating an average: total score divided by the number of items (scores).\n","    average_score = total_score / scores_number  \n","    \n","    # I round the calculated average to two decimal places using the round() function\n","    average_score = round(average_score, 2)  \n","    \n","    # I return a dictionary containing the student's name and their rounded average score.\n","    return {'name': student['name'], 'average_score': average_score}\n","\n","# This function checks if a student has an average score greater than 90. \n","# I will use this function later in the filter() function to filter out students who don’t meet the condition of having an average score above 90\n","def above_90(student):\n","    \n","    # I return True if the student's average score is greater than 90, and False otherwise.\n","    # This is the condition that I’ll use to filter out students. Only those who meet this condition will be included in the final list.\n","    return student['average_score'] > 90\n","\n","# This function organizes the process of calculating the average scores and filtering the students. \n","def above90_students(students):\n","    \n","    # I use the map() function to apply the \"AverageCalculation\" function to each student in the \"students\" list. \n","    # The result is converted to a list using list(). -> map() returns an iterator, so I need to convert it to a list to work with it later.\n","    student_avg_scores = list(map(AverageCalculation, students))  \n","    \n","    # I use the filter() function to apply the \"above_90\" function to the list student_avg_scores, returning only students whose average score is above 90.\n","    stud_above90 = filter(above_90, student_avg_scores)\n","    \n","    # Since filter() returns an iterator, I need to convert it to a list so I can return it in the correct format.\n","    return list(stud_above90)\n","\n","# This function generates a list of students with random names and random scores.\n","def generate_random_student_dataset(num_students=50):\n","    \n","    # I create a list of student names in the format Student 1, Student 2, ..., up to the specified number of students (num_students)\n","    names = [f\"Student {i}\" for i in range(1, num_students + 1)]\n","    \n","    # I create a list of dictionaries where each dictionary represents a student with a name and a list of random scores.\n","    # The number of scores is also random, between 3 and 6.\n","    dataset = [\n","        {\n","            \"name\": name,\n","            \"scores\": [randint(50, 100) for _ in range(randint(3, 6))]  # Punteggi casuali tra 50 e 100\n","        }\n","        for name in names\n","    ]\n","    \n","    # I return the generated list of students.\n","    return dataset\n","\n","# I generate a random dataset with 50 students.\n","random_student_dataset = generate_random_student_dataset(50)\n","\n","# I pass the random student dataset to the above90_students function, which calculates the average scores and filters out the students with average scores > 90.\n","result = above90_students(random_student_dataset)\n","\n","# I print the final list of students whose average score is greater than 90.\n","for element in result:\n","   print(f\"{element} \\n\")\n"]},{"cell_type":"markdown","metadata":{"id":"hWL_3xWNpkzj"},"source":["## Exercise 2\n","\n","You have a list of dictionaries, each representing a product with the following properties: name, price, and category. Using the functions `map`, `filter`, and `reduce`, calculate the average price of the products in each category and return a list of dictionaries containing only the categories where the average price exceeds 50.\n","\n","Example input:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"wwk7f8Ihpkzk"},"outputs":[],"source":["products = [\n","    {\"name\": \"Product A\", \"price\": 60, \"category\": \"Electronics\"},\n","    {\"name\": \"Product B\", \"price\": 40, \"category\": \"Electronics\"},\n","    {\"name\": \"Product C\", \"price\": 70, \"category\": \"Home\"},\n","    {\"name\": \"Product D\", \"price\": 30, \"category\": \"Home\"},\n","    {\"name\": \"Product E\", \"price\": 90, \"category\": \"Sports\"}\n","]"]},{"cell_type":"markdown","metadata":{"id":"Agz3cP7Ppkzl"},"source":["Use `map`, `reduce` and `filter` that produce an output like:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"LbFtUV_apkzl","outputId":"745cdcbe-7320-4a08-de7b-0e1bcb84473e"},"outputs":[{"data":{"text/plain":["[{'category': 'Electronics', 'average_price': 50.0},\n"," {'category': 'Sports', 'average_price': 90.0}]"]},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["[\n","    {\"category\": \"Electronics\", \"average_price\": 50.0},\n","    {\"category\": \"Sports\", \"average_price\": 90.0}\n","]"]},{"cell_type":"markdown","metadata":{"id":"nKBEAQE3pkzl"},"source":["### Test\n","Test your solution using the dataset generated by the following function."]},{"cell_type":"code","execution_count":4,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":284,"status":"ok","timestamp":1732181933353,"user":{"displayName":"Marco Zecchini","userId":"04080461015443912025"},"user_tz":-60},"id":"0qj_9nZSpkzm","outputId":"7bab35c7-84d0-4603-a1da-dc4a76768179"},"outputs":[{"data":{"text/plain":["[{'name': 'Product 1', 'price': 112, 'category': 'Home'},\n"," {'name': 'Product 2', 'price': 139, 'category': 'Clothing'},\n"," {'name': 'Product 3', 'price': 168, 'category': 'Electronics'},\n"," {'name': 'Product 4', 'price': 137, 'category': 'Books'},\n"," {'name': 'Product 5', 'price': 97, 'category': 'Clothing'}]"]},"execution_count":4,"metadata":{},"output_type":"execute_result"}],"source":["def generate_random_product_dataset(num_products=100):\n","    categories = [\"Electronics\", \"Home\", \"Sports\", \"Books\", \"Clothing\", \"Toys\"]\n","    dataset = [\n","        {\n","            \"name\": f\"Product {i}\",\n","            \"price\": randint(10, 200),  # Random price between 10 and 200\n","            \"category\": choice(categories),  # Randomly choose a category\n","        }\n","        for i in range(1, num_products + 1)\n","    ]\n","    return dataset\n","\n","# Example of using the function\n","random_dataset = generate_random_product_dataset(100)\n","random_dataset[:5]  # Display the first 5 entries to check the dataset structure\n"]},{"cell_type":"code","execution_count":29,"metadata":{"id":"AG9V3Wt7pkzm"},"outputs":[{"name":"stdout","output_type":"stream","text":["{'category': 'Electronics', 'average_price': 101.68}\n","{'category': 'Books', 'average_price': 112.7}\n","{'category': 'Toys', 'average_price': 127.38}\n","{'category': 'Home', 'average_price': 106.0}\n","{'category': 'Sports', 'average_price': 90.0}\n","{'category': 'Clothing', 'average_price': 112.68}\n"]}],"source":["# your code goes here\n","# hints: 1) Group products by category (you don't need to use map reduce for this part), then 2) use map reduce paradigm to\n","# calculate the average price for each category and filter categories with an average price > 50\n","\n","# I import randint and choice from the random module to generate random data for product prices and categories.\n","from random import randint, choice\n","\n","# I import reduce from the functools module, which will help me calculate the sum of prices in a list of products.\n","from functools import reduce\n","\n","# I define a function that takes a list of products (in the same category) as input and calculates the average price for the products in that list.\n","def AveragePrice(products):\n","    \n","    # I use reduce to sum up the price of each product in the list\n","    # The lambda x, y: x + y['price'] is a function that adds the price from each dictionary (y) in the list to the accumulator (x), starting from 0.\n","    total_price = reduce(lambda x, y: x + y['price'], products, 0)\n","    \n","    # After getting the total price, I calculate the average by dividing the total price by the length of the list (len(products))\n","    average_price = total_price / len(products)\n","    \n","    # I then return a dictionary containing the category (from the first product in the list) and the average_price, rounded to 2 decimal places using \"round\"\n","    return {\"category\": products[0]['category'], \"average_price\": round(average_price, 2)}\n","\n","# I define a function that groups products by their category\n","def GroupbyCategory(products):\n","    \n","    # I create an empty dictionary \"grouped\" to hold the products organized by their categories.\n","    grouped = {}\n","    \n","    # I loop through each product in the list: \n","    for product in products:\n","        \n","        # I check if the category of the current product is not already in the grouped dictionary:\n","        if product['category'] not in grouped:\n","            \n","            # If the category is not in the dictionary, I add an entry for that category with an empty list.\n","            grouped[product['category']] = []\n","        \n","        # I append the current product to the list corresponding to its category in the grouped dictionary.\n","        grouped[product['category']].append(product)\n","    \n","    # After processing all the products, I return the grouped dictionary, which contains categories as keys and lists of products as values.\n","    return grouped\n","\n","# I define a function that filters out categories where the average_price is less than or equal to 50.\n","def FilterCategories(categories):\n","    \n","    # I use the filter function with a lambda function to iterate through each dictionary in the categories list\n","    # The lambda checks if the average_price of each category is greater than 50: If it is, the category is kept; otherwise, it is discarded.\n","    # The filter function returns an iterator, so I will later convert it into a list.\n","    return filter(lambda x: x['average_price'] > 50, categories)\n","\n","# I define a function, which processes the list of products to filter out categories with average prices greater than 50.\n","def AvgPrices_above50(products):\n","    \n","    # I call GroupbyCategory to group the products by their category and store the result in grouped_products\n","    grouped_products = GroupbyCategory(products)\n","    \n","    # I use map to apply the \"AveragePrice\" function to each group of products (i.e., each category).\n","    # This produces a list of dictionaries where each dictionary contains a category and its average price.\n","    category_avg_prices = list(map(AveragePrice, grouped_products.values()))\n","    \n","    # I pass the result of map to \"FilterCategories\" to filter out categories where the average price is not greater than 50.\n","    filtered_categories = list(FilterCategories(category_avg_prices))\n","    \n","    # I return the filtered list of categories with their average prices above 50.\n","    return filtered_categories\n","\n","# I define a function to generate a random list of products.\n","# The function takes a parameter num_products to determine how many products to generate.\n","def generate_random_product_dataset(num_products=100):\n","    \n","    # I create a list of categories (Electronics, Home, Sports, etc.), and for each product, I randomly assign a price (using randint(10, 200)) and a category (using choice(categories) to pick randomly).\n","    categories = [\"Electronics\", \"Home\", \"Sports\", \"Books\", \"Clothing\", \"Toys\"]\n","    \n","    # I use a list comprehension to create a list of dictionaries, where each dictionary represents a product\n","    # As said, each product has a random price between 10 and 200, and a random category chosen from the categories list.\n","    dataset = [\n","        {\n","            \"name\": f\"Product {i}\",\n","            \"price\": randint(10, 200),  # Random price between 10 and 200\n","            \"category\": choice(categories),  # Randomly choose a category\n","        }\n","        for i in range(1, num_products + 1)\n","    ]\n","    \n","    # I return the generated dataset.\n","    return dataset\n","\n","# I call generate_random_product_dataset with the argument 100 to generate a dataset of 100 random products.\n","# I store the result in random_dataset.\n","random_dataset = generate_random_product_dataset(100)\n","\n","# I call the \"AvgPrices_above50\" function with the random_dataset to process the products: grouping them by category, calculating their average prices, \n","# and filtering out categories with average prices greater than 50.\n","# I store the result in the variable \"result\".\n","result = AvgPrices_above50(random_dataset)\n","\n","# I use a for loop to iterate over each item (category) in the result list.\n","# For each category, I print it on a new line, which results in each dictionary being displayed separately.\n","for category in result:\n","    print(category)\n"]},{"cell_type":"markdown","metadata":{"id":"hivtZEf7pkzm"},"source":["# Exercise 3\n","\n","You have a list of dictionaries, each representing an employee with the following properties: name, salary, and department. Your task is to use `map`, `filter`, and `reduce` to calculate the average salary for each department and return a list of dictionaries containing only the departments where the average salary is above 65,000.\n","\n","**Example Input**"]},{"cell_type":"code","execution_count":5,"metadata":{"executionInfo":{"elapsed":313,"status":"ok","timestamp":1732181936402,"user":{"displayName":"Marco Zecchini","userId":"04080461015443912025"},"user_tz":-60},"id":"N8vjLRHxpkzm"},"outputs":[],"source":["employees = [\n","    {\"name\": \"John\", \"salary\": 70000, \"department\": \"Engineering\"},\n","    {\"name\": \"Jane\", \"salary\": 75000, \"department\": \"Engineering\"},\n","    {\"name\": \"Alice\", \"salary\": 60000, \"department\": \"HR\"},\n","    {\"name\": \"Bob\", \"salary\": 68000, \"department\": \"HR\"},\n","    {\"name\": \"Charlie\", \"salary\": 90000, \"department\": \"Marketing\"},\n","    {\"name\": \"Diana\", \"salary\": 50000, \"department\": \"Marketing\"}\n","]"]},{"cell_type":"markdown","metadata":{"id":"otSniMO7pkzm"},"source":["Use `map`, `reduce` and `filter` that produce an output like:"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Kx4HT8RXpkzn","outputId":"8653ff13-815d-4040-c68e-fc4a6825134d"},"outputs":[{"data":{"text/plain":["[{'department': 'Engineering', 'average_salary': 72500.0},\n"," {'department': 'Marketing', 'average_salary': 70000.0}]"]},"execution_count":16,"metadata":{},"output_type":"execute_result"}],"source":["[\n","    {\"department\": \"Engineering\", \"average_salary\": 72500.0},\n","    {\"department\": \"Marketing\", \"average_salary\": 70000.0}\n","]"]},{"cell_type":"markdown","metadata":{"id":"GD_xlB78pkzn"},"source":["### Test\n","\n","Test your solution using the dataset generated by the following function."]},{"cell_type":"code","execution_count":6,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":508,"status":"ok","timestamp":1732181939215,"user":{"displayName":"Marco Zecchini","userId":"04080461015443912025"},"user_tz":-60},"id":"RhR9JLK-pkzn","outputId":"72bc934d-4d3c-477e-cf84-3cf1d8fae61a"},"outputs":[{"data":{"text/plain":["[{'name': 'Employee 1', 'salary': 40737, 'department': 'Engineering'},\n"," {'name': 'Employee 2', 'salary': 77275, 'department': 'Engineering'},\n"," {'name': 'Employee 3', 'salary': 107774, 'department': 'IT'}]"]},"execution_count":6,"metadata":{},"output_type":"execute_result"}],"source":["def generate_random_employee_dataset(num_employees=50):\n","    departments = [\"Engineering\", \"HR\", \"Marketing\", \"Sales\", \"Finance\", \"IT\"]\n","    dataset = [\n","        {\n","            \"name\": f\"Employee {i}\",\n","            \"salary\": randint(40000, 120000),  # Random salary between 40,000 and 120,000\n","            \"department\": choice(departments)  # Randomly choose a department\n","        }\n","        for i in range(1, num_employees + 1)\n","    ]\n","    return dataset\n","\n","random_employee_dataset = generate_random_employee_dataset(50)\n","\n","random_employee_dataset[:3]  # Display the first 3 entries of each dataset for checking\n"]},{"cell_type":"code","execution_count":30,"metadata":{"id":"pt9m6NK-pkzo"},"outputs":[{"name":"stdout","output_type":"stream","text":["{'department': 'Finance', 'average_salary': 77205.0}\n","{'department': 'Engineering', 'average_salary': 83699.5}\n","{'department': 'Sales', 'average_salary': 79646.8}\n","{'department': 'Marketing', 'average_salary': 75670.0}\n","{'department': 'IT', 'average_salary': 81591.82}\n","{'department': 'HR', 'average_salary': 76998.25}\n"]}],"source":["# your code goes here\n","# hints: 1) Group employees' salaries by department (you don't need to use map reduce for this part), then 2) use map reduce paradigm to\n","# calculate the average salary for each department and filter departments with an average salary > threshold\n","\n","# I import randint to generate random numbers for salaries -> It allows me to define a range (ex: 40,000 to 120,000) and pick numbers randomly within that range.\n","# I import choice to randomly assign a department to each employee from a predefined list of departments.\n","from random import randint, choice\n","\n","# I import reduce from functools. This function helps me repeatedly apply an operation (like summing salaries) across a sequence of data.\n","# I use it to calculate the total salary for employees in each department.\n","from functools import reduce\n","\n","# I define this function to calculate the average salary for a specific department. The input is a list of employees, each represented as a dictionary.\n","def AverageSalary(employees):\n","    \n","    # I use reduce to iterate over the list of employees and sum their salaries:\n","    # - The lambda function takes two arguments: x (the current total) and y (the current employee dictionary).\n","    # - For each employee, I extract the salary field and add it to x\n","    # - The starting value of x is 0.\n","    total_salary = reduce(lambda x, y: x + y['salary'], employees, 0)\n","    \n","    # I divide the total salary by the number of employees in the department to get the average.\n","    average_salary = total_salary / len(employees)\n","    \n","    # I create a dictionary containing the department name (extracted from the first employee in the list) and the average salary.\n","    # I use \"round\" to limit the average salary to 2 decimal places for clarity.\n","    return {\"department\": employees[0]['department'], \"average_salary\": round(average_salary, 2)}\n","\n","# I define this function to group employees by their departments.\n","# The input is a list of employee dictionaries.\n","def GROUPbyDEPARTMENT(employees):\n","    \n","    # I create an empty dictionary to hold departments names as keys and lists of employees (in that department) as values.\n","    grouped = {}\n","    \n","    # I loop through each employee in the input list.\n","    for employee in employees:\n","        \n","        # If the employee's department doesn’t already exist in the dictionary, ..\n","        if employee['department'] not in grouped:\n","            \n","            # .. -> I add it as a key and initialize its value to an empty list.\n","            grouped[employee['department']] = []\n","        \n","        # I append the current employee to the list corresponding to their department.\n","        grouped[employee['department']].append(employee)\n","    \n","    # I return the dictionary, which groups all employees by their department.\n","    return grouped\n","\n","# I define this function to filter out departments with an average salary of 65,000 or less.\n","def FilterDepartments(departments):\n","    \n","    # I use the filter function to keep only the departments where the average salary exceeds 65,000:\n","    # - I use the filter function to iterate over the list of department dictionaries\n","    # - For each department, the lambda function checks if the average_salary is greater than 65,000.\n","    # - The function returns an iterator containing only the departments that pass this condition.\n","    return filter(lambda x: x['average_salary'] > 65000, departments)\n","\n","def departments_above_salary_threshold(employees):\n","    \n","    # I call GROUPbyDEPARTMENT to organize employees into groups based on their department.\n","    grouped_employees = GROUPbyDEPARTMENT(employees)\n","    \n","    # I use map to apply AverageSalary to each group of employees (values of the grouped dictionary).\n","    # This generates a list of dictionaries, each containing the department name and its average salary.\n","    department_avg_salaries = list(map(AverageSalary, grouped_employees.values()))\n","    \n","    # I use FilterDepartments to keep only the departments with an average salary above 65,000.\n","    filtered_departments = list(FilterDepartments(department_avg_salaries))\n","    \n","    # I return the filtered list of departments.\n","    return filtered_departments\n","\n","# I define this function to generate a random dataset of employees.\n","def generate_random_employee_dataset(num_employees=50):\n","    \n","    # I list the possible departments to which employees can belong.\n","    departments = [\"Engineering\", \"HR\", \"Marketing\", \"Sales\", \"Finance\", \"IT\"]\n","    \n","    # I use a list comprehension to create a list of dictionaries for \"num_employees\" employees.\n","    # Each dictionary includes: - A unique name (ex:\"Employee 1\"); - A randomly generated salary between 40,000 and 120,000 using randint;\n","    # - A randomly assigned department using choice \n","    dataset = [\n","        {\n","            \"name\": f\"Employee {i}\",\n","            \"salary\": randint(40000, 120000),\n","            \"department\": choice(departments),\n","        }\n","        for i in range(1, num_employees + 1)\n","    ]\n","    \n","    # I return the list of generated employee dictionaries.\n","    return dataset\n","\n","# I generate a random dataset of 50 employees.\n","random_employee_dataset = generate_random_employee_dataset(50)\n","\n","# I call \"departments_above_salary_threshold\" to calculate average salaries by department and filter based on the salary threshold.\n","result = departments_above_salary_threshold(random_employee_dataset)\n","\n","# I iterate over the filtered list of departments and print each department's name and average salary.\n","for department in result:\n","    print(department)\n"]},{"cell_type":"markdown","metadata":{"id":"bzxr0v2Rpkzo"},"source":["# Biopython\n","\n","Write the following five functions to analyze global alignments between two sequences using Biopython's `pairwise2` module:\n","\n","1. **countMatches(s1, s2)**  \n","   This function takes two sequences (`s1`, `s2`) aligned using global alignment (pairwise2.globalxx) of the same length. It returns the number of positions where the elements of both sequences match.\n","\n","2. **countMismatches(s1, s2)**  \n","   This function takes two sequences (`s1`, `s2`) aligned using global alignment of the same length. It returns the number of positions where the elements of the two sequences are different (i.e., they are not gaps, and the characters do not match).\n","\n","3. **countGapOpens(s1, s2)**  \n","   This function takes two sequences (`s1`, `s2`) aligned using global alignment of the same length. It returns the number of gap openings in the alignment (a gap is opened when a '-' appears in the sequence).\n","\n","4. **countGapExtensions(s1, s2)**  \n","   This function takes two sequences (`s1`, `s2`) aligned using global alignment of the same length. It returns the number of gap extensions (where '-' continues in the alignment after an initial gap is opened).\n","\n","5. **getScore(s1, s2, matchScore, mismatchPenalty, gapOpenPenalty, gapExtensionPenalty)**  \n","   This function takes two sequences (`s1`, `s2`) aligned using global alignment and returns the alignment score based on the provided scoring scheme: `matchScore` for matches, `mismatchPenalty` for mismatches, `gapOpenPenalty` for opening a gap, and `gapExtensionPenalty` for extending a gap."]},{"cell_type":"code","execution_count":31,"metadata":{"id":"5PIaf6BUpkzo"},"outputs":[{"name":"stdout","output_type":"stream","text":["Calculated Score: 103.0\n","Pairwise2 Score: 103.0\n"]}],"source":["# Add your functions here\n","\n","# I import the pairwise2 module from Biopython, which provides tools for pairwise sequence alignments, both local and global.\n","from Bio import pairwise2\n","\n","# I import the SeqIO module from the Biopython library -> it helps handling sequence input and output\n","from Bio import SeqIO\n","\n","# I define this function to count matching positions between s1 and s2 where neither character is a gap ('-'):\n","# - Using zip, I pair corresponding characters from the two sequences.\n","# - I use a generator expression to count cases where the characters are identical (a == b) and not gaps (a != '-')\n","# - I return the total count of matches.\n","def countMatches(s1, s2):\n","\n","    # zip(s1, s2) -> The zip function takes two sequences (in this case, s1 and s2, which are strings) and pairs up the corresponding elements from each sequence into tuples\n","    # So, for each position in both strings, zip combines the characters at that position into a tuple.\n","\n","    # for a, b in zip(s1, s2) -> This part is a generator expression that loops over each tuple produced by zip; \n","    # It unpacks each tuple into the variables a and b, where a is the character from s1 and b is the character from s2\n","\n","    # if a == b and a != '-' -> this condition checks 2 things: a == b -> This checks if the characters from s1 and s2 are the same at the current position.\n","    # a != '-' -> This ensures that the character \"a\" is not a gap (-); This is important because we don't want to count matches where one or both sequences have gaps.\n","    \n","    # If both conditions (a == b and a != '-') are met, the expression returns 1 for that pair of characters.\n","    # So this counts how many times the characters in s1 and s2 are the same and neither of them is a gap.\n","\n","    # sum(...) -> The sum function takes the generator expression (the sequence of 1s) and adds them together.\n","    # so it gives the total count of matches between s1 and s2, excluding gaps.\n","    return sum(1 for a, b in zip(s1, s2) if a == b and a != '-')\n","\n","\n","# I use this function to count positions where characters differ (excluding gap characters)\n","# This function returns the mismatch count\n","def countMismatches(s1, s2):\n","    \n","    # I iterate over paired characters and count cases where they differ (a != b) but are not gaps (a != '-' and b != '-')\n","    return sum(1 for a, b in zip(s1, s2) if a != b and a != '-' and b != '-')\n","\n","\n","# I use this function to count the number of new gaps introduced in the alignment.\n","def countGapOpens(s1, s2):\n","    gap_opens = 0\n","    \n","    # For each position i ..\n","    for i in range(len(s1)):\n","        \n","        # .. I check if there is a gap in either sequence (s1[i] == '-' or s2[i] == '-').\n","        # A gap is counted as an \"opening\" if it is at the start of the sequence (i == 0) or follows a non-gap position (s1[i - 1] != '-' and s2[i - 1] != '-').\n","        if (s1[i] == '-' or s2[i] == '-') and (i == 0 or s1[i - 1] != '-' and s2[i - 1] != '-'):\n","            \n","            # I increment gap_opens for each new gap opening\n","            gap_opens += 1\n","    \n","    # I return the total\n","    return gap_opens\n","\n","\n","# I use this function to count positions where a gap continues beyond its opening.\n","def countGapExtensions(s1, s2):\n","    \n","    gap_extensions = 0\n","    \n","    # Starting from position i = 1, ..\n","    for i in range(1, len(s1)):\n","        \n","        # .. I check if there is a gap (s1[i] == '-' or s2[i] == '-'\n","        # and If the previous position also contains a gap (s1[i - 1] == '-' or s2[i - 1] == '-')..\n","        if (s1[i] == '-' or s2[i] == '-') and (s1[i - 1] == '-' or s2[i - 1] == '-'):\n","            \n","            # ..  it is counted as an extension.\n","            gap_extensions += 1\n","    \n","    # The function returns the total number of gap extensions.\n","    return gap_extensions\n","\n","\n","# In this function, I'm bringing all the previous functions together: I calculate matches, mismatches, gap opens, and gap extensions -> in order to compute a score for a global alignment between the two sequences (s1 and s2)\n","# Each of these factors is weighted by predefined scoring parameters\n","# The function takes 6 arguments: \n","# 1/2)s1, s2: The two aligned sequences for comparison; 3)matchScore: The score to be added for each match between s1 and s2; \n","# 4)mismatchPenalty: The penalty to subtract for each mismatch; 5)gapOpenPenalty: The penalty to subtract for opening a gap (= when a gap is introduced into the sequence alignment).\n","# 6)gapExtensionPenalty: The penalty to subtract for extending an existing gap (i.e., when a gap continues in the alignment).\n","def getScore(s1, s2, matchScore, mismatchPenalty, gapOpenPenalty, gapExtensionPenalty):\n","    \n","    # I call the previously defined functions to calculate the relevant quantities:\n","\n","    # This returns the number of positions where s1 and s2 match, excluding gaps.\n","    matches = countMatches(s1, s2)\n","    \n","    # This returns the number of positions where s1 and s2 are mismatched (both characters are not equal and not gaps)\n","    mismatches = countMismatches(s1, s2)\n","    \n","    # This returns the number of times a gap is opened in the alignment.\n","    gap_opens = countGapOpens(s1, s2)\n","    \n","    # This returns the number of times a gap is extended (i.e., it continues after the first gap is opened).\n","    gap_extensions = countGapExtensions(s1, s2)\n","    \n","    # I calculate the score by multiplying the number of matches by the matchScore, the number of mismatches by the mismatchPenalty, \n","    # the number of gap openings by the gapOpenPenalty, and the number of gap extensions by the gapExtensionPenalty\n","    score = (\n","\n","        # The formula then adds or subtracts these weighted values: - Positive for matches (because a match increases the score);\n","        # - Negative for mismatches, gap openings, and gap extensions (because these reduce the alignment quality).\n","        matches * matchScore -\n","        mismatches * mismatchPenalty -\n","        gap_opens * gapOpenPenalty -\n","        gap_extensions * gapExtensionPenalty\n","    )\n","    \n","    # I return the computed score.\n","    return score\n","\n","\n","# TEST\n","def main():\n","    \n","    # SeqIO.read reads the fasta file and returns an object SeqRecord\n","    # The sequence is accessible via .seq\n","    s1_record = SeqIO.read(\"./biopython-datasets/IL12A.fasta\", \"fasta\")\n","    s2_record = SeqIO.read(\"./biopython-datasets/IL12B.fasta\", \"fasta\")\n","    \n","    # To work with the sequence, I convert Seq into a string with str()\n","    s1 = str(s1_record.seq)\n","    s2 = str(s2_record.seq)\n","    \n","    # I define scoring parameters: - matchScore: Reward for matching characters; mismatchPenalty: Penalty for mismatched characters; \n","    # gapOpenPenalty: Penalty for introducing a new gap; gapExtensionPenalty: Penalty for continuing a gap\n","    matchScore = 1\n","    mismatchPenalty = 0\n","    gapOpenPenalty = 0\n","    gapExtensionPenalty = 0\n","    \n","    # I use pairwise2.align.globalms to perform a global alignment of s1 and s2\n","    alignments = pairwise2.align.globalms(\n","        s1, s2, \n","        \n","        # I pass the scoring parameters, converting penalties into negative values.\n","        # This generates a list of possible alignments.\n","        matchScore, -mismatchPenalty, -gapOpenPenalty, -gapExtensionPenalty\n","    )\n","    \n","    # I select the first alignment from the list (alignments[0]) and unpack the relevant details:\n","    # - aligned_s1 and aligned_s2: Aligned versions of the sequences.\n","    # - pairwise2_score: The alignment score calculated by pairwise2.\n","    aligned_s1, aligned_s2, pairwise2_score, _, _ = alignments[0]\n","    \n","    # I calculate the alignment score using the getScore function\n","    calculated_score = getScore(\n","        aligned_s1, aligned_s2, \n","        matchScore, mismatchPenalty, gapOpenPenalty, gapExtensionPenalty\n","    )\n","    \n","    # I use :.1f in an f-string in order to format the number as a floating-point number with one decimal place. \n","    # This means that even if the number is an integer, it will be displayed with one decimal, ensuring uniformity in the output format.\n","    print(f\"Calculated Score: {calculated_score:.1f}\")\n","    print(f\"Pairwise2 Score: {pairwise2_score:.1f}\")\n","\n","# I call the main function to execute the program.\n","main()\n","\n"]},{"cell_type":"markdown","metadata":{"id":"81O19v1fpkzo"},"source":["### Test\n","Align the sequences of the [Interleukin-12](https://en.wikipedia.org/wiki/Interleukin_12) chain A (denoted as `s1`) from the file [`IL12A.fasta`](https://qcbsciprolab2020.readthedocs.io/en/latest/file_samples/IL12A.fasta) and the Interleukin-12 chain B (denoted as `s2`) from the file [`IL12B.fasta`](https://qcbsciprolab2020.readthedocs.io/en/latest/file_samples/IL12B.fasta) and check the score as computed from pairwise2 and from your functions."]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ZBhfr3jepkzp"},"outputs":[],"source":["# add the output of the test here\n","\n","\"\"\"\"\"\n","\n","Calculated Score: 103.0\n","Pairwise2 Score: 103.0\n","\n","\"\"\"\"\""]}],"metadata":{"colab":{"provenance":[]},"kernelspec":{"display_name":"venv2","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.12.7"}},"nbformat":4,"nbformat_minor":0}
